### DISCUSSION TITLE
Add custom attributes to the <body> tag

### DISCUSSION THREAD

**Discussion by mimshins** [vercel/next.js#12325] [Help]
How can I (for example) add a custom className to `body` dynamically.
...or change the “dir” attribute value.

**Comment by c0b41** [comment]


**Reply by arsnl** [reply]
As the question mentioned to "add a custom className to tag **dynamically**", I don't think the `_document` is the good place for that. Given that `_document` is not hydrated client-side, `_app` or any lower page/component's are more adequate for dynamic values.

See my answer bellow: https://github.com/vercel/next.js/discussions/12325#discussioncomment-484771

**Comment by ParallelUniv3rse** [comment]
Any hints on how to do what the op asked? I'm having the same problem and simply modifying the _document.js will not work to change the <body> attributes per page(route), since 

> Document is only rendered in the server, event handlers like onClick won't work

As mentioned in the caveats.

**Comment by gdenisov** [comment]
I have the same question. react-helmet has 'bodyAttributes' and 'htmlAttributes' that are very useful for getting the right attributes in place per route in SSR. Is there a way to tap into this with the next/head?

**Comment by srikanth912** [comment]
How about using [withRouter](https://nextjs.org/docs/api-reference/next/router#withrouter) and then extracting the current route/path to do the comparison and inject properties based on the current router/path? 

`const { router: { pathname, query, asPath }` = this.props ?

**Comment by gdenisov** [comment]
@srikanth912 That's a good trick! The downside to this is that the logic is now in multiple places — while the rest of the page meta info is in the <Head> component.

**Comment by ghost** [comment]
This seems like a missing feature, no? It seems like refactoring from `react-helmet` should be relatively straightforward 🤔.

**Comment by maickoldxd** [comment]
if you use something like a wrapper, you can use componentDidMount

`class Container extends React.Component{

    componentDidMount() {
        const body = document.getElementsByTagName('body')[0];
        //do some stuff...
    }
    render(){
        {this.props.children}
     }
}`

**Comment by muuvmuuv** [comment]
Can you make this an issue not a discussion? It really IS a missing feature. I will continue using react-helmet-async instead because I am setting classes on body to hide certain elements on the page based on the route.

**Comment by arsnl** [comment]


**Reply by tadeolinco** [reply]
@arshad does this approach create this warning https://nextjs.org/docs/messages/next-head-count-missing if we put a `body` tag inside `Head`?

**Reply by jpedroschmitz** [reply]
At least for me, this doesn't work. It messes up the entire HTML and adds all meta tags inside the body and not on the `head`.

**Reply by vidhanio** [reply]
Thank you, this worked like a charm for me! :smile:

**Reply by ItsDaveB** [reply]
> @arshad does this approach create this warning https://nextjs.org/docs/messages/next-head-count-missing if we put a `body` tag inside `Head`?

It did for me!

**Comment by smnh** [comment]
@mimshins 

For best results, use both `_document.js` and `_app.js` to ensure that your server side rendered pages have the correct body class (or any other attribute on body or html) and that these classes updated correctly when user navigates between pages.

```jsx
// src/pages/_document.js

import Document, { Html, Head, Main, NextScript } from 'next/document';

export default class MyDocument extends Document {
    render() {
        const pageProps = this.props?.__NEXT_DATA__?.props?.pageProps;
        return (
            <Html>
                <Head />
                <body className={ pageProps.isDark ? 'dark-mode' : 'light-mode' }>
                    <Main />
                    <NextScript />
                </body>
            </Html>
        );
    }
}
```

```jsx
// src/pages/_app.js

import React, { useEffect } from 'react';

export default function MyApp({ Component, pageProps }) {
    useEffect(() => {
        document.body.className = pageProps.isDark ? 'dark-mode' : 'light-mode';
    });
    return <Component {...pageProps} />
}
```

```jsx
// src/pages/index.js

export async function getStaticProps() {
    return { props: { isDark: true } };
}

export default function IndexPage(props) {
    return (
        <div>
            <h1>Home Page</h1>
        </div>
    );
}
```

Here is the [post with full explanation](http://smnh.me/add-class-to-body-tag-in-nextjs/).

**Reply by wdcs-urmilpatel** [reply]
great work

**Reply by mimshins** [reply]
Great workaround ^^

**Reply by whale2002** [reply]
thanks

**Reply by ZB333ZB** [reply]
That's great. I wonder how can I use it with _redux_. I want to put `"overflow-hidden"` (Tailwind) on the body tag when a burger button is enabled. I reach the burger state via `useSelector`. It is `true` or `false`. I pass the value into the props as String (I get error if I don't)
```
//index.js

let isBurgerEnabled;

export default function Home() {
  isBurgerEnabled = useSelector(burgerState);
   /* ************************************* */
}

export async function getStaticProps() {
  return { props: { isOverflowHidden: String(isBurgerEnabled) } };
}

```
```
//_document.js
import Document, { Html, Head, Main, NextScript } from "next/document";

export default class MyDocument extends Document {
  render() {
    const pageProps = this.props?.__NEXT_DATA__?.props?.pageProps;
    return (
      <Html>
        <Head />
        <body
          className={`${
            pageProps.isOverflowHidden ? "overflow-hidden" : "overflow-auto"
          }`}
        >
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}
```
```
//_app.js
import React, { useEffect } from "react";
import { Provider } from "react-redux";
import { store } from "/src/redux/store";
import "/src/shared/styles/tailwind.css";

export default function MyApp({ Component, pageProps }) {
  useEffect(() => {
    document.body.className = pageProps.isOverflowHidden
      ? "overflow-hidden"
      : "overflow-auto";
  });

  return (
    <Provider store={store}>
      <Component {...pageProps} />
    </Provider>
  );
}
```

But I get false everytime. If I hardcode `false` or `true` it works fine
